### ISSUE TITLE
sftp GET: Unable to send FXP_CLOSE command

### ISSUE THREAD

**Issue by doublex** [libssh2/libssh2#360]
Current git master fails to load a file via SFTP. Error:
````Unable to send FXP_CLOSE command````

https://github.com/libssh2/libssh2/commit/4ac1ba71813308993a7c413853bc866eb6a92410 works
https://github.com/libssh2/libssh2/commit/78fdc41c56ca29fe8d4ead22d8140aeccdebaa7e fails

I have tried to investigate, but `git` confuses me:
````
git clone https://github.com/libssh2/libssh2.git a
git clone https://github.com/libssh2/libssh2.git b
cd a
git reset --hard 4ac1ba71813308993a7c413853bc866eb6a92410
cd ../b
git reset --hard 78fdc41c56ca29fe8d4ead22d8140aeccdebaa7e
cd ..
kdiff3 a b
````

How could this patch https://github.com/libssh2/libssh2/commit/78fdc41c56ca29fe8d4ead22d8140aeccdebaa7e change so many files?

`git log` does not report any changes between https://github.com/libssh2/libssh2/commit/4ac1ba71813308993a7c413853bc866eb6a92410 and https://github.com/libssh2/libssh2/commit/78fdc41c56ca29fe8d4ead22d8140aeccdebaa7e 

@willco007
Maybe I misunderstood git?

**Comment by willco007** [comment]
The two commits you've linked to have no functional differences, something else is going on.

**Comment by filom** [comment]
I believe the problem are cause earlier with libssh2_sftp_read when suddenly return LIBSSH2_ERROR_OUT_OF_BOUNDARY. 
Then the subsequent libssh2_sftp_close return an error too.

**Comment by willco007** [comment]
@filom Any chance you could run this down a bit more? I don't have an FXP project handy to debug this in. Thanks!

**Comment by doublex** [comment]
I disagree.
There is an issue with `git` (maybe even a security issue).

There are two commits from `fullermd`:
https://github.com/libssh2/libssh2/commits?author=fullermd

Both patches pretend to only change a typo: `OPENSSL_HEADER_BEGIN -> OPENSSH_HEADER_BEGIN`

Steps to test:
````
# First patch
git show d47af1bfc7ecaa7226ce9d30483e02363f8827d2 -1
# Second patch
git show 78fdc41c56ca29fe8d4ead22d8140aeccdebaa7e -1
````

But both patches do a lot of changes (the below one fixes the typo: `OPENSSL -> OPENSSH`)
````
# First patch (diff to previous patch)
git diff 09cc9aba28fbde29041e3790d6b0e03edb2588e8 d47af1bfc7ecaa7226ce9d30483e02363f8827d2
# Second patch (diff to previous patch)
git diff 4ac1ba71813308993a7c413853bc866eb6a92410 78fdc41c56ca29fe8d4ead22d8140aeccdebaa7e
````

Maybe they are special crafted pull requests.

**Comment by msohn** [comment]
09cc9aba28fbde29041e3790d6b0e03edb2588e8 is not a predecessor of d47af1bfc7ecaa7226ce9d30483e02363f8827d2 but it's the third predecessor of 78fdc41c56ca29fe8d4ead22d8140aeccdebaa7e

4ac1ba71813308993a7c413853bc866eb6a92410 is the second commit after d47af1bfc7ecaa7226ce9d30483e02363f8827d2 so why are you comparing it to 78fdc41c56ca29fe8d4ead22d8140aeccdebaa7e

I recommend you run
$ gitk --all
on your repository and have a look at the version graph between the commits you are referring to in this bug.

**Comment by knittl** [comment]
Those two commits do not have a direct parent-child relation, but there are many commits inbetween. `git log` shows all of them and shows many changes between those two commits:

```
$ git log --oneline 4ac1ba7..78fdc41
78fdc41 Spell OpenSS_H_ right when talking about their specific private key (#321)
9ae4b9e Fix more scope and printf warning errors
01c67b2 Silence unused variable warning
09cc9ab Fix errors identified by the build process
b242bde Fix casting errors after merge
...
```

`-p` would show you the patch/changes for each commit. You could also use gitk and inspect each commit visually: `gitk 4ac1ba7..78fdc41`

**Comment by doublex** [comment]
SFTP "GET" enters this line - introduced here: https://github.com/libssh2/libssh2/commit/634444a34899b8353f1959d73cf3594569e53b97
https://github.com/libssh2/libssh2/blob/master/src/transport.c#L572

**Comment by doublex** [comment]
I can't reproduce this issue with GDB.

```
if(numbytes <= (int)(total_num - (p->wptr - p->payload))) {
    memcpy(p->wptr, &p->buf[p->readidx], numbytes);
}
else {
    fprintf(stderr, "%d %d %d\n", numbytes, total_num, p->wptr - p->payload );
    return LIBSSH2_ERROR_OUT_OF_BOUNDARY;
}
````
writes:
````
# First run
24 -2123676427 15899
24 3 15899

# Second run
32 -576371467 15899
32 3 15899

# Third run
32 -521276171 15899
32 3 15899
````

**Comment by filom** [comment]
> @filom Any chance you could run this down a bit more? I don't have an FXP project handy to debug this in. Thanks!

I noticed the problem when I use a buffer bigger than 4096 on `libssh2_sftp_read`

**Comment by doublex** [comment]
@filom 
Try this patch:
https://github.com/libssh2/libssh2/pull/363

**Comment by filom** [comment]
> @filom
> Try this patch:
> #363

It works! Great!

**Comment by doublex** [comment]
The problem is, that `size_t total_num` is used uninitialized.

**Comment by doublex** [comment]
Replacing `total_num` with `p->total_num` should be the fix

**Comment by willco007** [comment]
Fixed with commit f5e84ae, thanks everyone for tracking this down and verifying it.
